### PR TITLE
Add `poll_for_reply` and `poll_for_reply_unchecked` functions

### LINKED ISSUE #244: Add `poll_for_reply` function
Supporting `xcb_poll_for_reply64` with a `poll_for_reply` function would allow asynchronous waiting for a number of replies.

I will give this a go myself.

### PR DESCRIPTION
**Pull request by Antikyth**
This PR adds `poll_for_reply` and `poll_for_reply_unchecked` functions - the async equivalents of the blocking `wait_for_reply` and `wait_for_reply_unchecked`. The poll API is based on the difference between `wait_for_event` and `poll_for_event`.

Where `wait_for_reply` and `wait_for_reply_unchecked` are the Rust bindings to `xcb_wait_for_reply64`, `poll_for_reply` and `poll_for_reply_unchecked` are the Rust bindings to `xcb_poll_for_reply64`.

Closes #244.

### CODE REVIEW

**Comment by Antikyth** [comment]
The changes to `wait_for_reply` and `wait_for_reply_unchecked` are just:
- Documentation changes
  - Fixed spelling mistake: "occured" -> "occurred"
  - Explained that they are the blocking versions, and linking to the non-blocking versions (the non-blocking versions do the same, linking to the blocking versions)
- Changing `std::ptr::null_mut()` to `ptr::null_mut()`, as `std::ptr` is imported and `ptr::null_mut()` is used in other places. It doesn't change the behavior at all.

File: src/base.rs [hunk 1]
```diff
@@ -1574,7 +1579,10 @@ impl Connection {
 
     /// Get the reply of a previous unchecked request.
     ///
-    /// If an error occured, `None` is returned and the error will be delivered to the event loop.
+    /// If an error occurred, `None` is returned and the error will be delivered to the event loop.
```
Comment (Antikyth) [inline src/base.rs:1582]: Fixed spelling mistake: "occured" -> "occurred".

File: src/base.rs [hunk 2]
```diff
@@ -1533,7 +1533,10 @@ impl Connection {
         self.check_request(self.send_request_checked(req))
     }
 
-    /// Get the reply of a previous request, or an error if one occured.
+    /// Gets the reply of a previous request, or an error if one occurred.
```
Comment (Antikyth) [inline src/base.rs:1536]: Fixed spelling mistake: "occured" -> "occurred".

File: src/ffi/ext.rs [hunk 3]
```diff
@@ -238,14 +238,14 @@ extern "C" {
...
-    ) -> *mut c_void;
+    ) -> *mut c_int;
 
     pub(crate) fn xcb_poll_for_reply64(
         c: *mut xcb_connection_t,
         request: u64,
         reply: *mut *mut c_void,
         e: *mut *mut xcb_generic_error_t,
-    ) -> *mut c_void;
+    ) -> *mut c_int;
```
Comment (Antikyth) [inline src/ffi/ext.rs:248]: `xcb_poll_for_reply` and `xcb_poll_for_reply64` return `c_int`, not `c_void`. The `c_void`, which represents the returned reply, is what `*reply` is set to. The return value instead represents whether the reply has been received yet: `0` if the reply has not been received (i.e. `None`), `1` if it has (i.e. `Some`).

File: src/base.rs [hunk 4]
```diff
@@ -1606,6 +1616,213 @@ impl Connection {
...
+    ///             wm_protocols_atom = conn
+    ///                 .poll_for_reply(&wm_protocols_cookie)?
+    ///                 .atom();
+    ///         }
+    ///         // If `wm_name_atom` is yet to be received, poll for it.
+    ///         if wm_name_atom.is_none() {
+    ///             wm_name_atom = conn
+    ///                 .poll_for_reply(&wm_name_cookie)?
+    ///                 .atom();
+	///         }
```
Comment (Antikyth) [inline src/base.rs:1657]:
```suggestion
    ///         }
```

File: src/base.rs [hunk 5]
```diff
@@ -1606,6 +1616,213 @@ impl Connection {
...
+    ///         mut wm_protocols_atom,
+    ///         mut wm_name_atom,
+    ///     ) = (None, None);
+    ///
+    ///     loop {
+    ///         // If `wm_protocols_atom` is yet to be received, poll for it.
+    ///         if wm_protocols_atom.is_none() {
+    ///             wm_protocols_atom = conn
+    ///                 .poll_for_reply(&wm_protocols_cookie)?
+    ///                 .atom();
```
Comment (Antikyth) [inline src/base.rs:1650]:
```suggestion
    ///                 .poll_for_reply(&wm_protocols_cookie)?;
```

File: src/base.rs [hunk 6]
```diff
@@ -1606,6 +1616,213 @@ impl Connection {
...
+    ///         }
+    ///         // If `wm_name_atom` is yet to be received, poll for it.
+    ///         if wm_name_atom.is_none() {
+    ///             wm_name_atom = conn
+    ///                 .poll_for_reply(&wm_name_cookie)?
+    ///                 .atom();
```
Comment (Antikyth) [inline src/base.rs:1656]:
```suggestion
    ///                 .poll_for_reply(&wm_name_cookie)?;
```

File: src/ffi/ext.rs (hunk 3 above)
Comment (rtbo) [reply src/ffi/ext.rs:248]: @Antikyth can you correct the return type of both FFI functions to `c_int` (as opposed to `*mut c_int`) ?
The return type is not a pointer in the C API.
https://gitlab.freedesktop.org/xorg/lib/libxcb/-/blob/master/src/xcbext.h?ref_type=heads#L291

Comment (Antikyth) [reply src/ffi/ext.rs]: > @Antikyth can you correct the return type of both FFI functions to `c_int` (as opposed to `*mut c_int`) ? The return type is not a pointer in the C API. https://gitlab.freedesktop.org/xorg/lib/libxcb/-/blob/master/src/xcbext.h?ref_type=heads#L291

Fixed it now (my apologies, I forgot about it - I wasn't able to when I saw the other message last week)

**Comment by rtbo** [comment]
@sprocklem is right, the return value of `xcb_poll_for_reply` indicates whether the request is finished. If request is finished we have to check if it is a reply or an error.

Therefore the return value of `Connection::poll_for_reply` should be `Option<Result<C::Reply>>` rather than `Result<Option<Reply>>`.
And to be consistent with `Connection::wait_for_reply_unchecked`, `Connection::poll_for_reply_unchecked` should return `Option<ConnResult<Option<C::Reply>>>`.

To summarize

| | checked | unchecked |
|-----|-----|------|
| blocking | `wait_for_reply -> Result<C::Reply>` | `wait_for_reply_unchecked -> ConnResult<Option<C::Reply>>` |
| non-blocking | `poll_for_reply -> Option<Result<C::Reply>>` | `poll_for_reply_unchecked -> Option<ConnResult<Option<C::Reply>>>` |

**Comment by rtbo** [comment]
Closed in favor of #248.
Thank you @Antikyth for bringing this up!
